xml_struct_derive/src/serialize.rs

```diff
@@ -0,0 +1,195 @@
+    generate_serialize_impl_for, with_enum_variants, with_struct_fields, with_text_variants, Field,
+    FieldKind, Variant, VariantKind,
+};
+
+/// Generates an implementation of the `XmlSerialize` trait for a Rust struct
+/// and its fields.
+pub(crate) fn write_serialize_impl_for_struct(
+    ident: Ident,
+    generics: Generics,
+    input: DataStruct,
+    props: TypeProps,
+) -> TokenStream {
+    let fields = match input.fields {
+        syn::Fields::Named(fields) => {
+            let map_result: Result<Vec<Field>, syn::Error> = fields
+                .named
+                .into_iter()
+                .map(|field| {
+                    // We should be able to unwrap without panicking, since we
+                    // know these are named fields.
+                    let ident = field.ident.unwrap();
+                    let accessor = quote!(self.#ident);
+
+                    let field = Field {
+                        kind: FieldKind::Named(ident),
+                        ty: field.ty.into_token_stream(),
+                        accessor,
+                        props: FieldProps::try_from_attrs(field.attrs, true)?,
+                    };
+
+                    Ok(field)
+                })
+                .collect();
+
+            match map_result {
+                Ok(fields) => fields,
+                Err(err) => return err.into_compile_error(),
+            }
+        }
+        syn::Fields::Unnamed(fields) => {
+            let map_result: Result<Vec<Field>, syn::Error> = fields
+                .unnamed
+                .into_iter()
+                .enumerate()
+                .map(|(idx, field)| {
+                    let idx_literal = Literal::usize_unsuffixed(idx);
+                    let accessor = quote!(self.#idx_literal);
+
+                    let field = Field {
+                        kind: FieldKind::Unnamed,
+                        ty: field.ty.into_token_stream(),
+                        accessor,
+                        props: FieldProps::try_from_attrs(field.attrs, false)?,
+                    };
+
+                    Ok(field)
+                })
+                .collect();
+
+            match map_result {
+                Ok(fields) => fields,
+                Err(err) => return err.into_compile_error(),
+            }
+        }
+        syn::Fields::Unit => vec![],
+    };
+
+    generate_serialize_impl_for(ident, generics, props, with_struct_fields(fields))
+}
+
+/// Generates an implementation of the `XmlSerialize` trait (and the
+/// `XmlSerializeAttr` trait if appropriate) for a Rust enum, its variants, and
+/// their fields.
+pub(crate) fn write_serialize_impl_for_enum(
+    ident: Ident,
+    generics: Generics,
+    input: DataEnum,
+    props: TypeProps,
+) -> TokenStream {
+    if props.should_serialize_as_text {
+        // We should already have verification that this enum consists solely of
+        // unit variants, so we just collect their identifiers.
```

Comment:
suggestion: Indicate where the verification takes place, e.g.:

```suggestion
        // We have already verified that this is a unit-only enum in
        // `TypeProps::try_from_input`, so we just collect the variants'
        // identifiers.
```

Reply:
I've tried to make this comment a little clearer, but in general I'm reluctant to refer to specific type/function names in comments since the type/function can change without any needed changes to this code, meaning this comment would fall out of date if there were a rename.